tests/prepare.ts

```diff
@@ -20,10 +20,10 @@ const defaultOptions: SwaggerConfig = {
     tsoa_auth: {
       authorizationUrl: 'http://swagger.io/api/oauth/dialog',
       flow: 'implicit',
-      scopes: {
+      /*scopes: {
         'read:pets': 'read things',
         'write:pets': 'modify things',
-      },
+      },*/
```

Comment:
I've never has this issue when compiling. So it's really surprising to me that you're having this issue. Perhaps you have the wrong stuff in your node_modules?

Can you please:
1) delete your node modules folder
2) install yarn globally
3) yarn install
4) revert this change (since I'm not sure what this breaks)
5) rerun the tests

Reply:
I believe this error is absolutely valid, look into: swagger/swagger.ts
  export interface OAuth2ImplicitSecurity {
    type: 'oauth2';
    description?: string;
    flow: 'implicit';
    authorizationUrl: string;
  }
OAuth2PasswordSecurity, OAuth2ApplicationSecurity and OAuth2AccessCodeSecurity have a scopes parameter, but OAuth2ImplicitSecurity simply does not have it... so quite unsure why this I'm the only one who gets this error? Maybe a newer typescript version shows this error (I'm on 3.7.2)

Reply:
I have reverted this change on my side because I think this should probably not be part of this PR, but the issue still exists on my side:
Error:(23, 7) TS2322: Type '{ api_key: { in: "query"; name: string; type: "apiKey"; }; tsoa_auth: { authorizationUrl: string; flow: "implicit"; scopes: { 'read:pets': string; 'write:pets': string; }; type: "oauth2"; }; }' is not assignable to type '{ [name: string]: Security; }'.
  Property 'tsoa_auth' is incompatible with index signature.
    Type '{ authorizationUrl: string; flow: "implicit"; scopes: { 'read:pets': string; 'write:pets': string; }; type: "oauth2"; }' is not assignable to type 'Security'.
      Object literal may only specify known properties, and 'scopes' does not exist in type 'OAuth2ImplicitSecurity'.

Reply:
so I can't explain the error or lack of error (because I'm on not on my laptop), but I can clarify the OAuth stuff.

Basically our security types are wrong. Check out the canonical swagger types that live in definitely typed. They show that scopes is optional is on all flows. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/swagger-schema-official/index.d.ts#L216

So basically, I'd leave prepare as it was before (I.e. Revert the change), but update our internal swagger type definitions to allow scopes to be passed for implicit type like it the [linked types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/swagger-schema-official/index.d.ts#L216) have it.

Reply:
fixed it over here: https://github.com/lukeautry/tsoa/pull/527/files